tensor_comprehensions/tc_unit.py

```diff
@@ -0,0 +1,674 @@
+        kwargs.update(self.kwargs)
+        name, backward_name = get_tc_names_from_kwargs(**kwargs)
+        kwargs.pop("name", None)
+        backward = True if backward_name is not None else False
+        hash_key = get_tc_hash_key(name, *input_tensors)
+
+        # lookup for the options in the cache. Whenever we make the call to
+        # autotune, tuning must happen. But if the kernel has been tuned earlier
+        # then we can use previous options to seed the tuning.
+        if hash_key in self.tuner_cache:
+            options_cache = self.tuner_cache[hash_key]
+        else:
+            options_cache = {}
+
+        # we give priority to the options user might have passed via file, or
+        # Options object.
+        cache_file = ""
+        if "cache" in kwargs and kwargs["cache"]:
+            if isinstance(kwargs["cache"], bool):
+                hash_key = get_tc_hash_key(name, *input_tensors)
+                cache_file = "/tmp/{}_{}".format(hash_key, str(uuid.uuid4()))
+            elif isinstance(kwargs["cache"], str):
+                cache_file = kwargs["cache"]
+            logger.info('Autotuning cache will be saved to: {}.cuda/options'.format(cache_file))
+        else:
+            logger.warning("Autotuning results won't be cached. 'cache' option is not set")
+
+        # we will first run the autotuning on the forward layer, the inputs are given
+        # for that, we will tune those
+        kwargs["type"] = "forward"
+        # we pass this tuner object so we can load from file without having to
+        # create special object
+        kwargs["tuner"] = self.autotuner
+        options = get_options_from_kwargs_and_tuner_cache(name, cache_file, options_cache, *input_tensors, **kwargs)
+        forward_best_options = self.tune_and_store(
+            name, input_tensors, mapping_options=options, cache_file=cache_file
+        )
+        # update the cache with the options
+        options_cache["forward"] = forward_best_options
+        if not backward:
+            self.tuner_cache[hash_key] = options_cache
+            return forward_best_options
+
+        # now, we have to tune the backward layer, for that, we need to run
+        # the forward layer first, get it's output,
+        logger.info('Autotuning the backward layer now')
+        cu = TcCompilationUnit()
+        cu.define(self.tc_lang)
+
+        if "options" in kwargs:
+            orig_options = kwargs["options"]
+            kwargs["options"] = forward_best_options
+            outputs = cu.compile_and_run(name, input_tensors, **kwargs)
+            kwargs["options"] = orig_options
+        else:
+            outputs = cu.compile_and_run(name, input_tensors, options=forward_best_options, **kwargs)
+        # now that we have the outputs of the forward pass, we have the inputs
+        # for the backward layer and we can now tune the backward layer
+        reorder_function = kwargs["reorder_function"] if "reorder_function" in kwargs else None
```

Comment:
Unless I am missing it in this large PR I do not see a test that uses a custom reorder_function.
Please add one such example. 
Anything simple that produces a temporary as part of the outputs is fine.

Reply:
added the example `test_python/layers/test_convolution_reorder.py`